### PR TITLE
Fixed for non-ascii characters and optimized

### PR DESCRIPTION
**Pull request by ericpauley**
This PR removes index access to input strings and instead uses range iteration and runeLen offsets.

It also switches from string representations to rune representations, and replaces toLower conversions with equalFold checks, which are significantly faster.

Performance numbers:
```
benchmark                                          old ns/op     new ns/op     delta
BenchmarkFind/with_unreal_4_(~16K_files)-8         18821066      4926121       -73.83%
BenchmarkFind/with_linux_kernel_(~60K_files)-8     45350582      11973487      -73.60%

benchmark                                          old allocs     new allocs     delta
BenchmarkFind/with_unreal_4_(~16K_files)-8         207238         897            -99.57%
BenchmarkFind/with_linux_kernel_(~60K_files)-8     190717         204            -99.89%

benchmark                                          old bytes     new bytes     delta
BenchmarkFind/with_unreal_4_(~16K_files)-8         5254408       158912        -96.98%
BenchmarkFind/with_linux_kernel_(~60K_files)-8     12121526      38496         -99.68%
```

### CODE REVIEW

**Comment by sahilm** [comment]
Wow! mind = blown. Please let me digest this code. I will merge later today. I'm probably going to come back and ask a few questions. I wouldn't want to pass up the opportunity to learn from you.

Thank you very much!

File: fuzzy.go [hunk 1]
```diff
@@ -55,35 +54,38 @@ characters upto the first match.
 */
 func Find(pattern string, data []string) []Match {
 	if len(pattern) == 0 {
-		return []Match{}
+		return nil
```
Comment (rmetzler) [inline fuzzy.go:57]: I guess it would be easier for client code if this just returned an empty array like before.

Comment (ichivack) [reply fuzzy.go:57]: In what case is an empty slice more usable than a nil slice? Note, none of the common things you do with a slice panics if the slice is nil ([see playground](https://play.golang.org/p/lgusiNOG3LJ)).

File: fuzzy.go [hunk 2]
```diff
@@ -124,45 +129,48 @@ func Find(pattern string, data []string) []Match {
...
 		}
 	}
-	return matches
-}
-
-func insertMatch(matches []Match, match Match) []Match {
-	i := sort.Search(len(matches), func(i int) bool {
-		return matches[i].score <= match.score
+	sort.SliceStable(matches, func(i int, j int) bool {
+		return matches[i].score >= matches[j].score
```
Comment (kovetskiy) [inline fuzzy.go:151]: seems like you changed the sort order

Comment (lunixbochs) [reply fuzzy.go:151]: The tests check the sort order.

File: fuzzy.go [hunk 3]
```diff
@@ -96,26 +98,29 @@ func Find(pattern string, data []string) []Match {
 					matchedIndex = j
 				}
 			}
-			nextp := ""
-			if patternIndex < len(pattern)-1 {
-				nextp = strings.ToLower(string(pattern[patternIndex+1]))
+			var nextp rune
```
Comment (lunixbochs) [inline fuzzy.go:106]: You can get ~12% speedup by reusing nextp as the next candidate

File: fuzzy.go [hunk 4]
```diff
@@ -55,35 +54,38 @@ characters upto the first match.
...
 				}
-				score += camelCaseBonus(j, candidate)
-				score += separatorBonus(j, candidate)
+				if unicode.IsLower(last) && unicode.IsUpper(candidate) {
+					score += camelCaseMatchBonus
+				}
+				if j != 0 && isSeparator(last) {
+					score += matchFollowingSeparatorBonus
+				}
 				if len(match.MatchedIndexes) > 0 {
```
Comment (lunixbochs) [inline fuzzy.go:94]: I don't think this index adjacency logic works with `j` being a byte index rather than a rune iterator. Perhaps keep a separate counter for index? `j` when ranging over a string is still the byte index, which breaks adjacency for multi-byte runes. See: https://play.golang.org/p/dVDNZyq6owZ

File: fuzzy.go (hunk 1 above)
Comment (sahilm) [reply fuzzy.go:57]: I have no strong preferences here. Go authors [prefer](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices) nil slices so let's go with returning a nil slice.

File: fuzzy.go (hunk 3 above)
Comment (sahilm) [reply fuzzy.go:106]: Nice! I'll leave @ericpauley to decide if they want to tackle this now. If not let's tackle this in another change :)

Thanks @lunixbochs.

File: fuzzy.go (hunk 4 above)
Comment (sahilm) [reply fuzzy.go:94]: Awesome catch! I guess the problem is that we don't have tests for Unicode.

@ericpauley Could you please fix this and add tests around Unicode handling (especially index tracking and adjacent match bonuses).

**Review by sahilm — CHANGES_REQUESTED** [review]
We should be good to go once we address [this](https://github.com/sahilm/fuzzy/pull/4#discussion_r184883650). Thanks!

File: fuzzy.go (hunk 4 above)
Comment (ericpauley) [reply fuzzy.go:94]: @lunixbochs Could you explain what the problem is here? Last isn't set using a byte index, it's set in the previous iteration, so multi-length runes should not be an issue. The only place this really needed to be considered was when calculating nextc, where a byte index was needed.

Comment (sahilm) [reply fuzzy.go:94]: I see a few problems with our Unicode handling. Let me add a few review notes for you.

Comment (ericpauley) [reply fuzzy.go:94]: Just realized what @lunixbochs meant and made some changes, should be good now as far as I can tell, also added a test case for multi-byte utf-8.

File: fuzzy.go (hunk 3 above)
Comment (ericpauley) [reply fuzzy.go:106]: I think you may have meant `nextc`. I modified the candidate traversal to take advantage of this, comes out to a 4-6% speedup all told.

File: fuzzy.go [hunk 5]
```diff
@@ -124,55 +134,59 @@ func Find(pattern string, data []string) []Match {
 					patternIndex++
 				}
 			}
+			lastIndex = j
+			last = candidate
 		}
 		// apply penalty for each unmatched character
 		penalty := len(match.MatchedIndexes) - len(data[i])
 		match.score += penalty
 		if len(match.MatchedIndexes) == len(pattern) {
```
Comment (sahilm) [inline fuzzy.go]: This comparison is invalid for Unicode. The following test fails -

```go
func TestFindWithUnicode(t *testing.T) {
	matches := Find("\U0001F41D", []string{"\U0001F41D"})
	if len(matches) != 1 {
		t.Errorf("got %v matches; expected 1 match", len(matches))
	}
}

```

**Review by sahilm — CHANGES_REQUESTED** [review]

File: fuzzy.go (hunk 5 above)
Comment (sahilm) [reply fuzzy.go]: @ericpauley Let me know if you'd like to collaborate on tests. If you give me edit access on the pull request I can try and write as many failing tests as I can.

**Comment by ericpauley** [comment]
@sahilm Thanks for finding that. I'm ready to hand off development on the unicode feature/tests. Feel free to merge into a feature branch and continue development.

File: fuzzy.go (hunk 3 above)
Comment (lunixbochs) [reply fuzzy.go:106]: Yep! I meant `nextc`, picked the wrong line when quickly looking for it in the github interface. And you're right, that wasn't the only tweak that helped for my 12%. This should be the other 6% (using an interface for sorting rather than a closure, and not making a duplicate call for `bonus`): https://github.com/lunixbochs/fuzzy/commit/23dec577d3cf95de1ba7d19c921a27122c209bac

**Comment by lunixbochs** [comment]
Kinda weird, because your Match objects are allocated as part of the array, which gets sorted, one more thing you'll get a perf bump from is reducing the size of your Match object to reduce memory copies. It has a few redundant fields.

You can remove Str (which is kind of an API choice, but it's redundant as you have Index) and score (score is private so it's only exposed to tests!).

(And at that memory size you might look into using a heap + pointer allocating your matches to skip the sort altogether)

File: fuzzy.go (hunk 3 above)
Comment (ericpauley) [reply fuzzy.go:106]: TIL, hadn't considered that the interface method would be faster but now it makes sense, no more reflection methods to access slice indices, just interface resolution for comparison. On the other hand it increases allocation, I wonder which one is more preferable. In the most performance-critical cases I generally reimplement sort on the concrete type. (Can't wait for Go 2 generics)

Comment (lunixbochs) [reply fuzzy.go:106]: How does it increase allocation? It's not really creating a new type or object, just associating methods to []Match at compile-time. Using an interface like this even seems to slightly reduce allocation in -benchmem for me (not allocating the sort closure).

**Review by sahilm — APPROVED** [review]

**Comment by sahilm** [comment]
Unicode handling seems to be A-ok. I haven't been able to find more bugs 😺  Thanks folks!

@lunixbochs I would like to apply some of your perf. optimisations. Would you like to make a pull request or should I go ahead steal your ideas? I'm fine either way :)

P.S. 'Twas awesome to see everyone come together and find/fix bugs. Given enough eyeballs all bugs **truly** are shallow.
